src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs

```diff
@@ -0,0 +1,146 @@
+﻿using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Corax;
+using Corax.Queries;
+using Raven.Server.Documents.Indexes.Static.Spatial;
+using Raven.Server.Documents.Queries;
+using Raven.Server.Documents.Queries.Results;
+using Raven.Server.Documents.Queries.Timings;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Utils;
+using Sparrow.Json;
+using Sparrow.Logging;
+using Voron.Impl;
+
+namespace Raven.Server.Documents.Indexes.Persistence.Corax
+{
+    public class CoraxIndexReadOperation : IndexReadOperationBase
+    {
+        private readonly IndexSearcher _indexSearcher;
+        private readonly CoraxQueryEvaluator _coraxQueryEvaluator;
+        private long _entriesCount = 0;
+        private const int _bufferSize = 2048;
+
+        public CoraxIndexReadOperation(Index index, Logger logger, Transaction readTransaction) : base(index, logger)
+        {
+            _indexSearcher = new IndexSearcher(readTransaction);
+            _coraxQueryEvaluator = new CoraxQueryEvaluator(_indexSearcher);
+        }
+
+        public override void Dispose()
+        {
+            _indexSearcher?.Dispose();
+        }
+
+        public override long EntriesCount() => _entriesCount;
+
+        public override IEnumerable<QueryResult> Query(IndexQueryServerSide query, QueryTimingsScope queryTimings, FieldsToFetch fieldsToFetch,
+            Reference<int> totalResults, Reference<int> skippedResults,
+            IQueryResultRetriever retriever, DocumentsOperationContext documentsContext, Func<string, SpatialField> getSpatialField, CancellationToken token)
+        {
+            var pageSize = query.PageSize;
+            var isDistinctCount = pageSize == 0 && query.Metadata.IsDistinct;
+            if (isDistinctCount)
+                pageSize = int.MaxValue;
+            var position = query.Start;
+            long readCounter = 0;
+
+            IQueryMatch result = _coraxQueryEvaluator.Search(query.Metadata, fieldsToFetch);
+            if (result == null)
+                yield break;
+
+            totalResults.Value = Convert.ToInt32(result.Count);
+            var ids = ArrayPool<long>.Shared.Rent(_bufferSize);
+            try
+            {
+                int read = 0;
+
+                //escaping already returned docs.
+                // Q: Can Corax have "skip" method inside? It's inefficient to copy values only for skipping.
```

Comment:
In most cases, we probably won't need that, but the issue is that the query may have a sort order, in which case we have to do the work anyway.
We can probably push the skip to the Corax code itself however. 
@redknightlois would that benefit us performance wise?

Reply:
Probably would make sense.